frontend/src/lib/components/modals/FormModal.svelte

```diff
@@ -1,40 +1,70 @@
-<script lang="ts">
-  import type { Readable } from 'svelte/store';
-
+<script context="module" lang="ts">
   import Modal, { DialogResponse } from '$lib/components/modals/Modal.svelte';
-  import { FormError, lexSuperForm } from '$lib/forms';
+  import type { LexFormState } from '$lib/forms/superforms';
+  import type { AnyZodObject, ZodObject, z } from 'zod';
+
+  export type FormModalResult<S extends AnyZodObject> = {
+    response: DialogResponse;
+    formState: LexFormState<S>;
+  };
+</script>
+
+<script lang="ts">
+  import { FormError, lexSuperForm, type ErrorMessage } from '$lib/forms';
   import Form from '$lib/forms/Form.svelte';
-  import type { ZodObject, z } from 'zod';
+  import type { Readable } from 'svelte/store';
 
   type Schema = $$Generic<ZodObject>;
   type FormType = z.infer<Schema>;
+  type SubmitCallback = (state: LexFormState<Schema>) => Promise<ErrorMessage>;
+
   export let schema: Schema;
 
   const superForm = lexSuperForm(schema, () => modal.submitModal());
-  const { errors, reset, message, enhance } = superForm;
-  const _form = superForm.form;
+  const { form: _form, errors, reset, message, enhance, formState } = superForm;
   let modal: Modal;
 
   export async function open(
-    onSubmit: (d: FormType) => Promise<string | undefined>,
-    value?: Partial<FormType>
-  ): Promise<void> {
+    value: Partial<FormType> | undefined,
+    onSubmit: SubmitCallback
+  ): Promise<FormModalResult<Schema>>;
+  export async function open(onSubmit: SubmitCallback): Promise<FormModalResult<Schema>>;
+  export async function open(
+    valueOrOnSubmit: Partial<FormType> | SubmitCallback | undefined,
+    _onSubmit?: SubmitCallback
+  ): Promise<FormModalResult<Schema>> {
+    const onSubmit = _onSubmit ?? (valueOrOnSubmit as SubmitCallback);
+    const value = _onSubmit ? (valueOrOnSubmit as Partial<FormType>) : undefined;
+
     if (value) _form.set(value, { taint: false });
-    if ((await modal.openModal()) === DialogResponse.Cancel) return;
-    const error = await onSubmit($_form);
-    if (error) {
-      $message = error;
-      // again go back to the top and await a response from the modal.
-      return await open(onSubmit);
-    }
+
+    const response = await openModal(() => onSubmit($formState));
```

Comment:
instead of wrapping our callback in a callback, can we just pass `onSubmit` directly into `openModal`, and just call it in `openModal` directly with the `$formState`? I feel like that might be easier to track the call stack, but I'm not 100% sure. What do you think?

Reply:
Yeah, I'm not sure why I did it the way I did, so I implemented your suggestion in develop here: https://github.com/sillsdev/languageforge-lexbox/commit/3933aacb9af0f3b1e807ef4a9766cbab8bfd75cd